### ISSUE TITLE
considering "License: BSD" as meaning bsd-new should include some notion of uncertainty

### ISSUE THREAD

**Issue by jonassmedegaard** [aboutcode-org/scancode-toolkit#1901]
The test src/licensedcode/data/rules/bsd-new_89.yml is categorized as bsd-new, but contains only this:
> License: BSD

I disagree that "BSD" equals "BSD-3-Clause" - sure, that is the "more original among the Free flavors" and also the more common among all of them.  But neither is a reason for reliable disambiguation.

I think the most sensible is to classify this as "any of the BSD licenses, but not possible to tell which".
In licensecheck that is labeled as "BSD~unspecified".

Consequently, I think test src/licensedcode/data/rules/bsd-new_100.RULE should be categorized as "BSD-3-Clause and/or BSD~unspecified" (not as "BSD-3-Clause" because even though the subsequent text is unambiguous, is it not clear if or how the two statements correlate (there is no implicit "one means same as the other" nor is there an "choose one or the other" or "you must apply both one and the other")

**Comment by pombredanne** [comment]
@jonassmedegaard you have a point but that's also something that we debated with several others including @royaljust @ariel11 and @kpfleming in particular as we curate things in the ClearlyDefined project... 

We have agreed to this so far:
https://github.com/clearlydefined/clearlydefined/pull/121/files#diff-850e84ef50dc7426c075c56ed961142cR97

> If the only license information is “BSD,” code the license as “BSD-3-Clause.” 

The rationale is that you are never wrong by treating this as “BSD-3-Clause” if this is a one or two clause BSD license in the sense that you are not omitting any of the three clauses.

@tieguy you had some opinion there too: your point of view would be welcomed 

FYI @jonassmedegaard maintains licensecheck @ Debian

**Comment by jonassmedegaard** [comment]
The rationale of saying BSD-3-Clause when in doubt is safe only within a world of _Free_ licenses.

How about non-free licenses, most famously known being the original BSD-4-Clause, but also newer problematic derivatives like SSLeay?

For someone sloppily refering to BSD, I would not be so certain that they chose only among 2 safe variants: The only thing I see as certain when stating "BSD" is that it is _not_ certain which it means.

**Comment by jonassmedegaard** [comment]
If it really is BSD-3-Clause, then how do I obey the third clause? See https://github.com/nexB/scancode-toolkit/issues/1900 for a concrete example of that dilemma.

**Comment by pombredanne** [comment]
@jonassmedegaard you wrote:
> How about non-free licenses, most famously known being the original BSD-4-Clause, but also newer problematic derivatives like SSLeay?

These BSD-4-Clause, SSLeay, OpenSSL, Apache-1/1.1 variants and more are all their own licenses so they are not at issue. They WILL be reported with their own license key. 

The BSD case you raised here and that I am interested is strictly when the only clue we have is `License: BSD`. Nothing else, no text, no other lead.


> For someone sloppily referring to BSD, I would not be so certain that they chose only among 2 safe variants: The only thing I see as certain when stating "BSD" is that it is _not_ certain which it means.

That's not certain for sure, but also generally I cannot see an issue with the BSD-3-Clause "maximum" approach taken here. Note that we could also ensure that for these cases we always have a lower rule `relevance` and therefore lower match score: that would capture well the uncertainty.

> If it really is BSD-3-Clause, then how do I obey the third clause? See #1900 for a concrete example of that dilemma.

The 3rd clause is the non-endorsement clause and AFAIK this is widely regarded as superfluous because endorsement and name usage are already protected anyhow, in particular by trademark laws. 
IMHO If you have a BSD-2-Clause you have no more right to name usage and  endorsement that you had in a BSD-3-Clause. 

So having the third clause helps to meet the condition to the full license being there but does not matter too much anyway.

To recap the point and argument I am making:
If someone says a program is `License BSD` then saying the text we should use is `BSD-3-Clause` because:
- it would cover BSD 0,1,2 and 3 that are what everyone in this field commonly refers to as the BSD licenses
- if it were something else such as OpenSSL or BSD original/4 Clauses then it would have been identified differently
- we/I should ensure that we attach a lower relevance to such a detection to properly signal to the user that there is uncertainty.

**Comment by pombredanne** [comment]
@spotrh @richardfontana I would be interested to get you take on this topic

**Comment by jonassmedegaard** [comment]
> The 3rd clause is the non-endorsement clause and AFAIK this is widely regarded as superfluous
clause being superfluous or not is besides the point of _detecting_ licenses.

Sorry, I see how my way of framing the question caused the derailing into discussing _use_ of licenses, so let me try clarify:

What I meant to ask is not "how can a recipient use a third clause with zero subjects?" but instead "how can a license scanner claim to have found a 3-clause BSD license when third clause is bogus - i.e. contains zero subjects?".

**Comment by richardfontana** [comment]
I largely agree with arguments of @pombredanne here. It's reasonable to treat "License: BSD" as meaning "BSD-3-Clause at the most restrictive". (This btw would obviously not be so if you somehow found "License: BSD" on some code from say 1997.) 

The fact that a standard license clause can be a no-op given the lack of information and specificity in the indication of licensing is not at all a new or surprising phenomenon in open source.

@jonassmedegaard  the advertising clause, while historically controversial, is not generally seen as non-FOSS. However the disfavor with which BSD-style advertising clauses have been viewed since the late 1990s is another reason to assume that an ambiguous "BSD" label in modern code does not encompass advertising clause-containing variants.

**Comment by jonassmedegaard** [comment]
thanks, @richardfontana, those are convincing arguments (and sorry if your points were the same, @pombredanne, and I simply didn't pay proper attention).

So what is left of this issue is the part of indicating uncertainty.

**Comment by pombredanne** [comment]
@jonassmedegaard would reporting these cases as "90%" relevant work for you?

**Comment by jonassmedegaard** [comment]
What would be helpful to me specifically is that e.g. "90%" is noted in the YAML data for the test cases.
(what Scancode toolkit is _reporting_ is less useful to me, because I don't actually run the tool, only reuse the test cases)

**Comment by pombredanne** [comment]
> What would be helpful to me specifically is that e.g. "90%" is noted in the YAML data for the test cases.

That would be possible, but the YAML test metadata only contain the list of matched licenses expressions today and there are no expectations of a certain score/coverage/relevance and that should be added (as an option)

**Comment by jonassmedegaard** [comment]
I understand that we are talking about a future extension to the test hints.
Including such probability note in cases known to be uncertain would be helpful - thanks!

**Comment by tieguy** [comment]
I'd still much prefer something like Licenseref-BSD-unknown so that the end-user can make an informed policy decision. I take Richard's point that 3-clause is almost always the reasonable answer in that scenario, but still seems to me that the scanner should not be (silently) making that policy assessment.

**Comment by DennisClark** [comment]
Nobody wants to spend time worrying about an ambiguously stated BSD license assignment, especially when a project might contain hundreds of such things.   Closing.

**Comment by jonassmedegaard** [comment]
> Nobody wants to spend time worrying about an ambiguously stated BSD license assignment, especially when a project might contain hundreds of such things. Closing.

So because noone likes uncertanty, this project chooses to hide discoveries of uncertainty. That is a form of user interface design choice, I guess, but it really smells funny to me!
